src/server.py

```diff
@@ -0,0 +1,188 @@
+    """
+    if databag is None:
+        return None
+    if relation_name == state.SLAVE_RELATION:
+        return Credentials.from_jenkins_slave_interface_dict(databag)
+    return Credentials.from_jenkins_agent_v0_interface_dict(databag, unit_name=unit_name)
+
+
+class ServerBaseError(Exception):
+    """Represents errors with interacting with Jenkins server."""
+
+
+class AgentJarDownloadError(ServerBaseError):
+    """Represents an error downloading agent JAR executable."""
+
+
+def download_jenkins_agent(server_url: str, connectable_container: ops.Container) -> None:
+    """Download Jenkins agent JAR executable from server.
+
+    Args:
+        server_url: The Jenkins server URL address.
+        connectable_container: The connectable agent container.
+
+    Raises:
+        AgentJarDownloadError: If an error occurred downloading the JAR executable.
+    """
+    try:
+        res = requests.get(f"{server_url}/jnlpJars/agent.jar", timeout=300)
+        res.raise_for_status()
+    except (requests.HTTPError, requests.Timeout, requests.ConnectionError) as exc:
+        logger.error("Failed to download agent JAR executable from server, %s", exc)
+        raise AgentJarDownloadError(
+            "Failed to download agent JAR executable from server."
+        ) from exc
+
+    connectable_container.push(
+        path=AGENT_JAR_PATH, make_dirs=True, source=res.content, user=USER, group=GROUP
+    )
+
+
+def validate_credentials(
+    agent_name: str, credentials: Credentials, connectable_container: ops.Container
+) -> bool:
+    """Check if the credentials can be used to register to the server.
+
+    Args:
+        agent_name: The Jenkins agent name.
+        credentials: Server credentials required to register to Jenkins server.
+        connectable_container: The Jenkins agent workload container.
+
+    Returns:
+        True if credentials and agent_name pairs are valid, False otherwise.
+    """
+    # IMPORTANT: add random jitter to prevent parallel execution.
+    # It's okay to use random since it's not used for sensitive data.
+    time.sleep(random.random())  # nosec
+    proc: ops.pebble.ExecProcess = connectable_container.exec(
+        [
+            "java",
+            "-jar",
+            str(AGENT_JAR_PATH),
+            "-jnlpUrl",
+            f"{credentials.address}/computer/{agent_name}/slave-agent.jnlp",
+            "-workDir",
+            str(JENKINS_WORKDIR),
+            "-noReconnect",
+            "-secret",
+            credentials.secret,
+        ],
+        timeout=5,
+        user=USER,
+        group=GROUP,
+        working_dir=str(JENKINS_WORKDIR),
+        combine_stderr=True,
+    )
+    # The process will exit due to connection failure(invalid credentials) or timeout.
+    # Check for successful connection log from the stdout.
+    connected = False
+    terminated = False
+    # The proc.stdout is iterable according to process.exec documentation
```

Comment:
What type does `mypy` think it is?

Reply:
Perhaps the issue is that it is optional? I.e., perhaps solvable using a `None` check?

Reply:
Otherwise we should give feedback to ops about this issue. @benhoyt FYI

Reply:
Thanks -- I've just tested, and yes, this is definitely an ops typing issue, so the `type: ignore` comment seems fair for now. I just released ops 2.4.0 with a bunch of improvements to the type hints, but unfortunately it doesn't fix this, sorry. I've opened an `ops` issue to track this case: https://github.com/canonical/operator/issues/963 ... static typing of I/O stuff in Python is super-annoying to get right. Go's `io.Reader` and `io.Writer` interfaces handle this so much better. Oh well. :-(

That said, it looks like this code is running a short-lived program and looping through it's output (presumably proc.stdout will reach EOF when the process exits? it should probably use `process.wait()` after the loop to tidy up). In that case, I'd strongly recommend using `proc.wait_output()`, as it's easier to get right and avoid race conditions; you would normally only iterate through `proc.stdout` directly when you need real-time streaming of the command's output. Something like this:

```python
proc = container.exec(..., combine_stderr=True)
stdout, _ = proc.wait_output()
for line in stdout.splitlines():
    ...
```

If, on the other hand, this actually is a long-running process, maybe it should be a Pebble service? But it doesn't look like it is.

Reply:
It is a long-running service that could fail, and just using time timeout exception to rely on whether the service has successfully connected wasn't quite reliable in this case, hence I needed to check the stream of outputs to see whether it has established connection successfully.
Thank you for your detailed explanation on the proc usage, I do miss Go's interfaces that could be much better here!